# Assignment 3 - Linear Classifiers

The goal of this assignment is to familiarize yourself with the SVM classifiers available in scikit-learn and practice applying them to data.

This assignment does _not_ require you to implement your own classifier from scratch, but you may need to look at the SciKit-Learn documentation to figure out how to call the library methods.

In [2]:
# import the things we'll need
import math
import numpy as np
from sklearn import neighbors
from sklearn import datasets
from sklearn import svm
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

## Load some data
Let's go ahead and use the iris dataset again.  Load it the same way we did in Assignment 1b, and split it into 60% train and 40% test just like we did there.

**Be sure to use the argument `random_state=0` just like we did in the previous assignment.**  This is a "seed" for the random number generator; any particular seed value should always result in the same set of "random" numbers.  We'll use fixed seed values here because it will allow everyone to get the same results.

In [130]:
# load the iris dataset and split it
iris = datasets.load_iris()
y=iris.target
# let's train a nearest neighbor classifier...
# note that we can use 'n_neighbors' to control the "K" in our KNN
classifierA = neighbors.KNeighborsClassifier(metric='euclidean', n_neighbors=1)
classifierA.fit(iris.data, iris.target)
#splitting our data into two parts 
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.4, random_state=1)

print("training shape:", X_train.shape, ", testing shape:", X_test.shape)

training shape: (90, 4) , testing shape: (60, 4)


## Create a Nearest Neighbor classifier
Create a nearest neighbor classifier just like we did in Assignment 1b; here, we will use it as a baseline for comparing our new classifiers.  Be sure to train it and then evaluate its performance on the testing data (again, you should be able to use code from the previous assignment with little or no modification to do this).  Use an n_neighbors value of 1.  This should give you the same accuracy as you got in the previous assignment.

Once you've got it working, try changing the `random_state` value in the train_test_split function to 1 and re-running your nearest-neighbor classifier.  Try it again with a value of 2.  Write down each of the accuracies below; the first one has been done for you.

In [131]:
# create, train, and test a nearest neighbor classifier
classifierB = neighbors.KNeighborsClassifier(metric='euclidean', n_neighbors=1)
classifierB.fit(X_train, y_train)
classifierB.score(X_test, y_test)

0.9666666666666667

**Question 1: accuracy of nearest neighbor for different random seed values**

seed=0, accuracy = 0.9166666666666666

seed=1, accuracy = 0.9666666666666667

seed=2, accuracy = 0.9666666666666667


## Create a linear SVM
Code to create a linear SVM is given; however, it is left to you to add lines to train the SVM on the training data and then evaluate its accuracy on the testing data.  This should work just like the training/testing process did using the nearestNeighbor class.

Test your classifier on different train/test splits using the same three random-seed values as previously, and report the accuracies you get.

In [132]:
linearSvm = svm.SVC(kernel='linear')
linearSvm.fit(X_train, y_train)
linearSvm.score(X_test, y_test)




# train and test

0.9833333333333333

**Question 2: accuracy of linear SVM for different random seed values**

seed=0, accuracy =0.9666666666666667

seed=1, accuracy =0.9833333333333333

seed=2, accuracy = 0.9666666666666667

## Create a polynomial-kernel SVM
Create another SVM, only this time use the argument `kernel='poly'` to make an SVM using a polynomial kernel.  Train and test it with different seed values as before.

In [133]:
# create, train, and test
linearSvm = svm.SVC(kernel='poly',C=1)
linearSvm.fit(X_train, y_train)
linearSvm.score(X_test, y_test)

0.9833333333333333

**Question 3: accuracy of polynomial SVM for different random seed values**

seed=0, accuracy =0.9666666666666667

seed=1, accuracy =0.9833333333333333

seed=2, accuracy = 0.95

## Create a rbf-kernel SVM

Create another svm, this time using `'rbf'` for the kernel type.  Train and test as before.

In [134]:
# create, train, and test
linearSvm = svm.SVC(kernel='rbf')
linearSvm.fit(X_train, y_train)
linearSvm.score(X_test, y_test)

0.9833333333333333

**Question 4: accuracy of rbf SVM for different random seed values**

seed=0, accuracy =0.95

seed=1, accuracy =0.9833333333333333

seed=2, accuracy =0.9666666666666667

In [3]:
math.sqrt(3)

1.7320508075688772